bigframes/core/compile/compiled.py

```diff
@@ -514,117 +643,621 @@ def aggregate(
+            self.predicates = list(predicates) if predicates is not None else None
+
+        def build(self) -> UnorderedIR:
+            return UnorderedIR(
+                table=self.table,
+                columns=self.columns,
+                predicates=self.predicates,
+            )
+
+
+class OrderedIR(BaseIbisIR, CompiledArrayValue):
+    """Immutable BigQuery DataFrames expression tree.
+
+    Note: Usage of this class is considered to be private and subject to change
+    at any time.
+
+    This class is a wrapper around Ibis expressions. Its purpose is to defer
+    Ibis projection operations to keep generated SQL small and correct when
+    mixing and matching columns from different versions of a DataFrame.
+
+    Args:
+        table: An Ibis table expression.
+        columns: Ibis value expressions that can be projected as columns.
+        hidden_ordering_columns: Ibis value expressions to store ordering.
+        ordering: An ordering property of the data frame.
+        predicates: A list of filters on the data frame.
+    """
+
+    def __init__(
+        self,
+        table: ibis_types.Table,
+        columns: Sequence[ibis_types.Value],
+        hidden_ordering_columns: Optional[Sequence[ibis_types.Value]] = None,
+        ordering: ExpressionOrdering = ExpressionOrdering(),
+        predicates: Optional[Collection[ibis_types.BooleanValue]] = None,
+    ):
+        super().__init__(table, columns, predicates)
+        # TODO: Validate ordering
+        if not ordering.total_ordering_columns:
+            raise ValueError("Must have total ordering defined by one or more columns")
+        self._ordering = ordering
+        # Meta columns store ordering, or other data that doesn't correspond to dataframe columns
+        self._hidden_ordering_columns = (
+            tuple(hidden_ordering_columns)
+            if hidden_ordering_columns is not None
+            else ()
+        )
+
+        # To allow for more efficient lookup by column name, create a
+        # dictionary mapping names to column values.
+        self._column_names = {column.get_name(): column for column in self._columns}
+        self._hidden_ordering_column_names = {
+            column.get_name(): column for column in self._hidden_ordering_columns
+        }
+        ### Validation
+        value_col_ids = self._column_names.keys()
+        hidden_col_ids = self._hidden_ordering_column_names.keys()
+
+        all_columns = value_col_ids | hidden_col_ids
+        ordering_valid = all(
+            col.column_id in all_columns for col in ordering.all_ordering_columns
+        )
+        if value_col_ids & hidden_col_ids:
+            raise ValueError(
+                f"Keys in both hidden and exposed list: {value_col_ids & hidden_col_ids}"
+            )
+        if not ordering_valid:
+            raise ValueError(f"Illegal ordering keys: {ordering.all_ordering_columns}")
+
+    @classmethod
+    def from_pandas(
```

Comment:
The separation for ordered/unordered here might be premature optimatization. Maybe we can share these implementations for now?

Reply:
Thats true, changed to always initialize as ordered version, and it will just drop the ordering property if in unordered compilation mode.